runtime/src/claims.rs

```diff
@@ -0,0 +1,278 @@
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+type EthereumAddress = [u8; 20];
+type EcdsaSignature = ([u8; 32], [u8; 32], i8);
+
+/// An event in this module.
+decl_event!(
+	pub enum Event<T> where
+		B = <T as balances::Trait>::Balance,
+		A = <T as system::Trait>::AccountId
+	{
+		/// Someone claimed some DOTs.
+		Claimed(A, EthereumAddress, B),
+	}
+);
+
+decl_storage! {
+	// A macro for the Storage trait, and its implementation, for this module.
+	// This allows for type-safe usage of the Substrate storage database, so you can
+	// keep things around between blocks.
+	trait Store for Module<T: Trait> as Claims {
+		Claims get(claims) build(|config: &GenesisConfig<T>| {
+			config.claims.iter().map(|(a, b)| (a.clone(), b.clone())).collect::<Vec<_>>()
+		}): map EthereumAddress => Option<T::Balance>;
+		Total get(total) build(|config: &GenesisConfig<T>| {
+			config.claims.iter().fold(Zero::zero(), |acc: T::Balance, &(_, n)| acc + n)
+		}): T::Balance;
+	}
+	add_extra_genesis {
+		config(claims): Vec<(EthereumAddress, T::Balance)>;
+	}
+}
+
+fn ecdsa_recover(sig: &EcdsaSignature, msg: &[u8; 32]) -> Option<[u8; 64]> {
+	let v = secp256k1::RecoveryId::parse(if sig.2 > 26 { sig.2 - 27 } else { sig.2 } as u8).ok()?;
+	let rs = (sig.0, sig.1).using_encoded(secp256k1::Signature::parse_slice).ok()?;
+	let pubkey = secp256k1::recover(&secp256k1::Message::parse(msg), &rs, &v).ok()?;
+	let mut res = [0u8; 64];
+	res.copy_from_slice(&pubkey.serialize()[1..65]);
+	Some(res)
+}
+
+fn create_msg(who: &[u8]) -> Vec<u8> {
+	let prefix = b"Pay DOTs to the Polkadot account:";
+	let mut l = prefix.len() + who.len();
+	let mut rev = Vec::new();
+	while l > 0 {
+		rev.push(b'0' + (l % 10) as u8);
+		l /= 10;
+	}
+	let mut v = b"\x19Ethereum Signed Message:\n".to_vec();
+	v.extend(rev.into_iter().rev());
+	v.extend_from_slice(&prefix[..]);
+	v.extend_from_slice(who);
+	v
+}
+
+fn eth_recover(s: &EcdsaSignature, who: &[u8]) -> Option<EthereumAddress> {
+	let msg = keccak256(&create_msg(who));
+	let mut res = EthereumAddress::default();
+	res.copy_from_slice(&keccak256(&ecdsa_recover(s, &msg)?[..])[12..]);
+	Some(res)
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event<T>() = default;
+
+		/// Make a claim.
+		fn claim(origin, ethereum_signature: EcdsaSignature) {
+			// This is a public call, so we ensure that the origin is some signed account.
+			let sender = ensure_signed(origin)?;
+			
+			let signer = sender.using_encoded(|data|
+					eth_recover(&ethereum_signature, data)
+				).ok_or("Invalid Ethereum signature")?;
+			
+			let balance_due = <Claims<T>>::take(&signer)
+				.ok_or("Ethereum address has no claim")?;
+			
+			<Total<T>>::mutate(|t| if *t < balance_due {
+				panic!("Logic error: Pot less than the total of claims!")
```

Comment:
This is more of a personal question, is there a reason to panic here versus just throwing an Err? Either way, if an error like this appears in the runtime, something is broken, but why not fail more gracefully?

Reply:
it already changed the state with the `take` above, so it can't return an error. panic shouldn't be a huge issue here since it doesn't do much before and it can only be abused if there's a critical logic error in the chain, which probably means there's a much bigger problem to worry about.